### PR TITLE
chore: use git hooks for linter/formatter and signed-off/conventional-commits

### LINKED ISSUE #2845: Automatically run linter and formatter as pre-commit hook
### Is your enhancement related to a problem? Please describe

Newcomers or existing contributors are forgetting to run formatter/linter then we need to amend our Pull Requests

### Describe the solution you'd like

- Run linter and formatter as pre-commit hook, thus when creating a PR it'll always be valid
- Include Signed-Off and check for conventional commits

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by benoitf**
### What does this PR do?
Add a pre-commit hook for linter/formatter
Add commit-msg hook for checking conventional commit and adding Signed-Off

### What issues does this PR fix or reference?

fixes https://github.com/containers/podman-desktop/issues/2845

### How to test this PR?

Change files (with unformatted style or lint errors)
Try to add the files to staging area and do a commit
it should fail

### CODE REVIEW

**Comment by odockal** [comment]
@benoitf Would be nice to add this information into CONTRIBUTING.md.

**Review by lstocchi — COMMENTED** [review]
It works but i don't like it much. For a simple commit it takes 10-24s. I would prefer to have a flag to skip the check when i think it's not necessary.

**Comment by benoitf** [comment]
@lstocchi 
`git commit -n` or `HUSKY=0 git <your-command>` would do what you expect

**Comment by benoitf** [comment]
if you think it will be a mess we could try to find other ways

**Review by lstocchi — APPROVED** [review]
@benoitf I didn't know the flag `-n` tbh. We can merge this and see how it goes.  
Everyone has his own way to work so if there is an option to customize the behavior, it's ok for me.

**Comment by benoitf** [comment]
I would say the main goal would be to have something self-explanatory for newcomers so their PR is "ready to go" without reading any documentation

Probably need to launch the unit tests as part of the hook maybe  (all tests except e2e)

**Review by mairin — APPROVED** [review]
LGTM

**Comment by jeffmaury** [comment]
Is it supposed to work on Windows ? I create a dummy commit with an additional empty file and WIP as commit message and got it accepted

**Comment by lstocchi** [comment]
> Is it supposed to work on Windows ? I create a dummy commit with an additional empty file and WIP as commit message and got it accepted

I tested it on Win 11 and it worked

**Comment by benoitf** [comment]
@jeffmaury did you run `yarn` or `yarn install` step ?

**Comment by jeffmaury** [comment]
Yes I did:

- pr checkout
- yarn install
- create file
- git add and git commit

**Comment by benoitf** [comment]
@jeffmaury tried on a Windows VM and it seems to work

I installed git with default/recommended options

https://github.com/containers/podman-desktop/assets/436777/ea3989b6-5abb-439f-8702-50e9b1c2e01a

**Review by cdrage — APPROVED** [review]

**Review by deboer-tim — APPROVED** [review]

**Comment by jeffmaury** [comment]
Seems this is caused by git-gui ignoring pre-commit hooks results. Works from CLI

**Comment by benoitf** [comment]
@jeffmaury is it helping following https://dev.to/studiospindle/using-husky-s-pre-commit-hook-with-a-gui-21ch ?

**Comment by jeffmaury** [comment]
That may not be an error but I noticed that from CLI an error is displayed but the commit has been done. Will check with a failing test

File: .husky/commit-msg [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+set -u
```
Comment (jeffmaury) [inline .husky/commit-msg:22]:
```suggestion
. "$(dirname -- "$0")/_/husky.sh"

set -u
```

**Review by jeffmaury — COMMENTED** [review]
Seems the husky bootstrap was missing. Now bad commit msg prevent commit be be done

**Review by jeffmaury — APPROVED** [review]
LGTM.

**Pull request merged**